nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/StandardCellFieldTypeReader.java

```diff
@@ -0,0 +1,129 @@
+ */
+class StandardCellFieldTypeReader implements CellFieldTypeReader {
+    private final TimeValueInference timeValueInference;
+
+    StandardCellFieldTypeReader(final TimeValueInference timeValueInference) {
+        this.timeValueInference = Objects.requireNonNull(timeValueInference, "Time Value Inference required");
+    }
+
+    /**
+     * Infer Cell Field Type and append possible Data Types to mapped Field Type Inference information
+     *
+     * @param cell Spreadsheet Cell required
+     * @param fieldName Cell field name for tracking in Field Types
+     * @param fieldTypes Map of Field Name to Field Type Inference information
+     */
+    @Override
+    public void inferCellFieldType(final Cell cell, final String fieldName, final Map<String, FieldTypeInference> fieldTypes) {
+        Objects.requireNonNull(cell, "Cell required");
+        Objects.requireNonNull(fieldName, "Field Name required");
+        Objects.requireNonNull(fieldTypes, "Field Types required");
+
+        final FieldTypeInference fieldTypeInference = fieldTypes.computeIfAbsent(fieldName, key -> new FieldTypeInference());
+        final DataType dataType = getCellDataType(cell);
+        fieldTypeInference.addPossibleDataType(dataType);
+    }
+
+    /**
+     * Get Record Data Type from Spreadsheet Cell Type and additional resolution of NUMERIC and STRING types
+     *
+     * @param cell Spreadsheet Cell required
+     * @return Record Data Type
+     */
+    @Override
+    public DataType getCellDataType(final Cell cell) {
+        Objects.requireNonNull(cell, "Cell required");
+
+        final CellType cellType = cell.getCellType();
+
+        final DataType dataType;
+
+        if (CellType.NUMERIC == cellType) {
+            // Date Formatting check limited to NUMERIC Cell Types
+            if (DateUtil.isCellDateFormatted(cell)) {
+                // Default to TIMESTAMP for date formatted values using standard Date Time Pattern
+                dataType = RecordFieldType.TIMESTAMP.getDataType();
+            } else {
+                final double numericCellValue = cell.getNumericCellValue();
+                final long roundedCellValue = (long) numericCellValue;
+                if (roundedCellValue == numericCellValue) {
+                    dataType = RecordFieldType.LONG.getDataType();
+                } else {
+                    // Default to DOUBLE for NUMERIC values following cell.getNumericCellValue()
+                    dataType = RecordFieldType.DOUBLE.getDataType();
+                }
+            }
+        } else if (CellType.BOOLEAN == cellType) {
+            dataType = RecordFieldType.BOOLEAN.getDataType();
+        } else if (CellType.STRING == cellType) {
+            final String cellValue = cell.getStringCellValue();
+            // Attempt Time Value inference for STRING cell values
+            final Optional<DataType> timeDataType = timeValueInference.getDataType(cellValue);
```

Comment:
That's a good question. It depends how flexible we should be for timestamp handling. In other words, it is possible for a `STRING` cell to contain something that looks like a timestamp. This would effectively remove the need for any time or date format properties in the `ExcelReader`, but it would require times and dates to be stored as `NUMERIC` types. I'm inclined to leave it in, given the existence of those format properties, but it might be helpful to keep it for now to enable wider time and date detection from some spreadsheets.

Reply:
I just stored a date in an Excel spreadsheet as a text e.g. 5/21/2024 so I see what you mean.

Reply:
I think though this piece should be documented that any time or date format properties  specified by the user (stored in `TimeValueInference`) will **_only_** be used to parse those cells whose cell type is string and not whose already is date.

Reply:
Thanks, I added some comments around the constructor.

Reply:
I actually meant end user documentation as this is a change in functionality from what has been up until now.

Reply:
Also it might be worth documenting that when inferring a schema, the schema will be limited to the following specific types BOOLEAN, DOUBLE,  LONG, STRING, TIMESTAMP and perhaps DATE and TIME.

Reply:
Reviewing the current `additionalDetails.md` documentation, the description actually seems to reflect the behavior accurately:

> If the "Time Format," "Timestamp Format," or "Date Format" properties are configured, any value that would otherwise be considered a STRING type is first checked against the configured formats to see if it matches any of them. If the value matches the Timestamp Format, the value is considered a Timestamp field. If it matches the Date Format, it is considered a Date field. If it matches the Time Format, it is considered a Time field. In the unlikely event that the value matches more than one of the configured formats, they will be matched in the order: Timestamp, Date, Time. I.e., if a value matched both the Timestamp Format and the Date Format, the type that is inferred will be Timestamp. Because parsing dates and times can be expensive, it is advisable not to configure these formats if dates, times, and timestamps are not expected, or if processing the data as a STRING is acceptable. For use cases when this is important, though, the inference engine is intelligent enough to optimize the parsing by first checking several very cheap conditions. For example, the string's length is examined to see if it is too long or too short to match the pattern. This results in far more efficient processing than would result if attempting to parse each string value as a timestamp.

I'm not sure how valuable it is to go into some additional implementation details at the moment, if future revisions are necessary. However, if you have a suggestion for an additional bullet point, glad to include.

Reply:
> any value that would otherwise be considered a STRING type
That part is bothering me as until now that was everything but now any column whose Excel format is date is not included in this. I am not sure users will understand that.

Reply:
I think part of the challenge is that Excel provides a narrow set of stored type values, and a slightly wider set of supported rendered type values. When combined with type inferencing, there is a lot of potential for confusion. That being said, I think the above paragraph is still accurate. At least the difference between a number of string should be clear enough, so that distinction remains, and is now more strongly followed.